src/ucp/tag/rndv.c

```diff
@@ -29,6 +29,88 @@ static int ucp_tag_rndv_is_get_put_op_possible(ucp_ep_h ep, ucp_lane_index_t lan
     }
 }
 
+static void ucp_tag_rndv_resolve_lanes(ucp_request_t *req)
+{
+    ucp_ep_h ep       = req->send.ep;
+    uint64_t lane_map = -1;
+    ucp_rndv_get_lane_info_t *info;
+    ucp_lane_index_t lane;
+    ucp_rsc_index_t i;
+
+    UCS_STATIC_ASSERT(UCP_MAX_LANES < (sizeof(lane_map) * 8));
+
+    ucs_assert(!ucp_ep_is_stub(ep));
+    ucs_assert_always(req->send.rndv_get.num_lanes <= UCP_MAX_RNDV_LANES);
+    ucs_assert(req->send.rndv_get.rkey != NULL);
+
+    /* already resolved? ok, just exit */
+    if (req->send.rndv_get.rkey->resolved) {
+        return;
+    }
+
+    /* lookup lanes for known memory domains */
+    for (i = 0; i < req->send.rndv_get.num_lanes; /* no increment here */) {
+        info = ucp_tag_rndv_get_lane_info(req, i);
+        ucs_assert(info->lane == UCP_NULL_LANE);
+        ucs_assert(info->md_index != UCP_NULL_RESOURCE);
+        for (lane = 0; lane < ucp_ep_num_lanes(ep); lane++) {
+            if ((lane_map & UCS_BIT(lane)) &&
+                (info->md_index == ucp_ep_md_index(ep, lane))) {
+                info->lane = lane;
+                lane_map &= ~UCS_BIT(lane); /* do not allow repeat lanes */
+                break;
+            }
+        }
+
+        if (info->lane != UCP_NULL_LANE) {
+            i++;
+        } else {
+            /* could not find pair lane - release rkey */
+            if (ucp_tag_rndv_is_rkey_valid(req, i)) {
+                uct_rkey_release(ucp_tag_rndv_rkey_bundle(req, i));
+            }
+            /* shift rest of rkeys to current location */
+            if (i + 1 < req->send.rndv_get.num_lanes) {
+                memmove(info, ucp_tag_rndv_get_lane_info(req, i + 1),
```

Comment:
maybe it's better to have a map of matched keys then memmove?

Reply:
bit optimized: now it copy last element to current position